pkg/lokifrontend/frontend/v2/frontend.go

```diff
@@ -144,19 +146,30 @@ func NewFrontend(cfg Config, ring ring.ReadRing, log log.Logger, reg prometheus.
 }
 
 func (f *Frontend) starting(ctx context.Context) error {
-	return errors.Wrap(services.StartAndAwaitRunning(ctx, f.schedulerWorkers), "failed to start frontend scheduler workers")
+	return errors.Wrap(services.StartAndAwaitRunning(context.Background(), f.schedulerWorkers), "failed to start frontend scheduler workers")
 }
 
-func (f *Frontend) stopping(_ error) error {
+func (f *Frontend) stopping(e error) error {
+	t := time.NewTicker(500 * time.Millisecond)
+	defer t.Stop()
+
+	start := time.Now()
+
+	for range t.C {
```

Comment:
I'm not a big fan of active polling. Ideally we would wait for done with a timeout and `select`. However, this is probably too complicated for this change.